### DISCUSSION TITLE
Manipulating the code generated by this utility

### DISCUSSION THREAD

**Discussion by TonyGravagno** [araera111/mysql-to-zod#49] [Show and tell]
I'd like to share what I'm doing with the code that comes from MySQL-to-Zod.

This utility puts the generated code into a single file. I need separate files for all of my schema definitions. I call every table an "entity". Each entity has its own schema file. It has a separate file for text messages like errors, labels, headers, etc. An entity has its own container page which then imports form components. Each of those form components is in a separate file, for data entry, inquiry, etc.

So a rough structure is as follows:

```
entities
   Airport
      AirportDb.ts
      AirportMaintForm.tsx
      AirportMaintPage.tsx
      AirportSchema.ts
      AirportText.ts 
   AuditLog
      AuditLogDb.ts
      AuditLogMaintForm.tsx
      AuditLogMaintPage.tsx
      AuditLogSchema.ts
      AuditLogText.ts 
```

<details>
<summary>This is the code that parses the one file into many files</summary>

```js
import fs from 'fs'
import path from 'path'
import yargs from 'yargs/yargs'
import { hideBin } from 'yargs/helpers'

// To execute: node genSource --target=..\schema

console.log('running genSource')
// Parse command line arguments
const argv = yargs(hideBin(process.argv)).argv
console.dir({ argv })
// Function to create the target directory if it doesn't exist
function createTargetDir(targetDir) {
  if (!fs.existsSync(targetDir)) {
    fs.mkdirSync(targetDir, { recursive: true })
  }
}

function main() {
 
  const source = `
  // entire file generated by MySQL-to-Zod
  // pasted here with no modification
  // because it does not include backticks.
  // I might enhance later to read the file
  // based on a CLI parameter 'source'
  `
  const targetDir = argv.target
  
  console.log( targetDir )

  createTargetDir(targetDir)
  
  console.log( 'dir created')

  const pattern = /export const ([A-Za-z0-9_]+) = z\.object\({[^]*?\}\);/g

  let match
  while ((match = pattern.exec(source)) !== null) {
    const [block, entityName] = match
    const targetCode = `//\n${block}\n`
    const targetFilePath = path.join(targetDir, `${entityName}.ts`)
    fs.writeFileSync(targetFilePath, targetCode, 'utf8')
    console.log(`File ${entityName}.ts has been created.`)
  }
}

main()

```

</details>

The resulting `${entityName}.ts` file is then put through multiple other parsing routines similar to this one, that look for key markers, extract relevant text, insert the text into a template, save the populated template for the current table, and then move on to the next table. Corresponding to the file names seen above these are the CLI commands to generate the files:

```
node genSchemaFile --source=..\schema --target=..\target
node genTextFile --source=..\schema --target=..\target
node genPageFile --source=..\schema --target=..\target
node genMaintFormFile --source=..\schema --target=..\target
node genDbFile --source=..\schema --target=..\target
```

Ideally, after all of these files are generated, there is full "referential integrity", where schema with error messages properly reference their corresponding text in the Text file, the MaintForms include a base input field for all columns, and the DbFile code has full CRUD for its table.

My current efforts relate to ensuring the proper form fields are generated based on the schema - Boolean becomes CheckBox, Date becomes a CalendarControl, etc. After that there is a manual process of fitting the form fields within a CSS Grid to position them in an attractive but responsive structure for data entry.

For me this becomes an ORM of sorts, and a dynamic form generator that can adapt to some degree to schema changes.

Very soon, I will need other features in MySQL-to-Zod (or a fork?) to add handling for contraints and foreign keys. This will allow client-side code to populate select boxes in one table's maintenance form with data from another table. It will also help to constrain fields, for example, so that birthdates cannot be beyond today or older than maybe 120 years.

The schema is already defined in the database. I want to use the same schema for server-side CRUD, for client-side forms, APIs, and any other interface with the application. Ideally I want to regenerate code across the application based on schema changes. From decades of experience I don't actually believe that's entirely possible, because small refinements will need to be made which will break the ability to replace live code with newly generated code. But I strive to get close to that ideal. This helps me to ensure that the code is developed in a highly modular fashion, so that I can regenerate as much as possible, and keep necessary post-generation customizations in separate modules. When I find that I'm customizing the same kind of code repeatedly, making further regenerations more difficult to re-integrate, I will attempt to further refactor and modularize out the original generated code.

Hey, this could be a total disaster later on, but when you're looking at creating a maintenance environment for a lot of tables, some code generation to kick off the project is a really attractive option. I'm just looking at 80 tables right now. I've heard others talk about hundreds. And this is only about fairly direct integrations with specific tables. In a real application, for example, we have an OrderForm which requires interfaces with Customers, OrderHeaders, OrderDetails, Inventory, and many other tables. It really helps to establish a consistent framework for as much of this as possible, so that we can spend more time on the stuff that cannot be generated.

**Comment by TonyGravagno** [comment]
BTW, with the recent enhancement to add comments into the generated code, I will use table comments to inject the full table name and some metadata into the generated code so that I can improve my CRUD code.

The field comments may be used to provide info about constraints because that's difficult for a utility like this to generate from SQL. I actually generate all of my SQL to CREATE TABLEs anyway. Now MySQL to Zod generates schema from those tables ... this is really cool. Since I have metadata about all fields, I can put that metadata into column comments, get those comments back in generated schema code, and then parse out those comments to generate constraints for schema and other application code.

As an example, a termination cannot come before a start date. The SQL column constraint syntax to do this is complex to parse, but easy to add as a comment.

Other business logic for table and column comments might include:

- Many record types should never be deleted, only flagged as inactive or archived.
- Is a field intended to be TINYINT(1) or Boolean?
- Is a specific field ever allowed to change after its first setting?
- Should a date field save the UTC timestamp, or a server-local date/time?

These are contrived examples and subject to debate. The real point is that metadata in comments can be used to help enforce business rules - until the SQL itself can be used for the same purpose.

I look forward to using this utility in future projects to create more solid foundations for new applications and to save more development time.

**Comment by araera111** [comment]
Thank you very much for taking the time to tell us about your project.
We hope mysql-to-zod can help you.

Your enthusiasm for automation sounds great.

Since I added the commenting feature, I thought it might be possible to work with AI (ex: github copilot) and others.
Generally, AI works better with appropriate comments.
This was another story.

Your idea is interesting and I am very grateful for your finding improvements.

As I mentioned before, the schema may be different for SELECT and INSERT.
Two schemas are needed for the same table.
I have not come up with a solution for this. Please let me know if you come up with something.

**Comment by TonyGravagno** [comment]
I don't have anything solid to offer at the moment, sorry.

I have not done anything with INSERT schema yet but I am aware of concerns. When using generated schema for a data entry form, I cannot include columns/fields that are auto-generated like primary keys or timestamps. For now I need to exclude those fields by looking for GENERATED in the CREATE TABLE ... but now with the comments included in schema this will be a lot easier for me to parse. Knowing the fields that should not be included for INSERT, I can generate new Types using Pick or Omit. Again, I have not done this yet but I know this is in my roadmap.

As an example, this is sample SQL from MySqlDump:

```sql
CREATE TABLE IF NOT EXISTS `Test_DB`.`schemaTesting` (
  `id` INT GENERATED ALWAYS AS () STORED COMMENT 'Generated PK',
  `name` VARCHAR(45) NULL COMMENT 'Simple text',
  `type` VARCHAR(2) NOT NULL DEFAULT 'AA' COMMENT 'ENUM:AA,BB,CC',
  UNIQUE INDEX `id_UNIQUE` (`id` ASC) VISIBLE,
  PRIMARY KEY (`id`))
ENGINE = InnoDB
COMMENT = 'For testing schema'

INSERT INTO `Test_DB`.`schemaTesting` (`id`, `name`, `type`) VALUES (DEFAULT, 'First', 'CC');
INSERT INTO `Test_DB`.`schemaTesting` (`id`, `name`, `type`) VALUES (DEFAULT, 'Second', 'BB');
```

Note that DEFAULT is used for a GENERATED field, and I added a comment that can be checked until the SQL is parsed for this detail. Note also how I've tagged `type` as an enum.

Other people have also [asked](https://stackoverflow.com/questions/3978326/get-insert-statement-for-existing-row-in-mysql) about how to get an INSERT statement. The method differs depending on the database. The code provided by people in that thread (translated to JS) may serve as a base in this package.

As an aside, for development projects and most production, I use MariaDB which has its own nuances and its own version of mysqldump = [mariadb-dump](https://mariadb.com/kb/en/mariadb-dump/). I also use MySQL itself, but not as often. I occasionally may reference some detail that is actually specific to MariaDB if I have not checked MySQL Docs. 😢

**Comment by TonyGravagno** [comment]
Because of recent improvements to this utility, my refactoring code now generates full forms with labels. All I need to do is to generate Zod from the DB, run my code, copy the results into my app, and the whole app shows all of the new menu items and the related forms when the menus are clicked. I giggled like a little kid when this worked the first time. 👦

Upcoming improvements include : 

- Single command to perform all operations above.
- Select type-specific components : Checkbox for Bool, Calendar, DateTime, Radio...
- Generate JSON using the DB data as a base, that has schema and other details specific to tables and columns, where such details will not come from from the DB.

**Comment by araera111** [comment]
Your automation is very interesting to me.

I never thought of automatically creating forms from a database.
That would save us a lot of work.

I would be happy to continue working with you.
If there are any new features you would like me to create, you are welcome to suggest them.
I am willing to take on the challenge.

**Reply by TonyGravagno** [reply]
I will clean up the code and share it with you in another repo. This will take some time, maybe a week because the code is undergoing significant change as I adapt to the new opportunities that come from what has been done here...
